### PR TITLE
graph: exported images times matching creation

### PR DESCRIPTION
**Pull request by vbatts**
the image export, that is used in `docker save` previous has just had
the layers times (atimes, mtimes) be when the save was done.

``` bash
vbatts@valse ~ (master) $ docker save busybox | tar tv
drwxr-xr-x 0/0               0 2015-09-03 22:22 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/
-rw-r--r-- 0/0               3 2015-09-03 22:22 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/VERSION
-rw-r--r-- 0/0            1405 2015-09-03 22:22 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/json
-rw-r--r-- 0/0         2643968 2015-09-03 22:22 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/layer.tar
drwxr-xr-x 0/0               0 2015-09-03 22:22 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/
-rw-r--r-- 0/0               3 2015-09-03 22:22 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/VERSION
-rw-r--r-- 0/0            1346 2015-09-03 22:22 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/json
-rw-r--r-- 0/0            1024 2015-09-03 22:22 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/layer.tar
drwxr-xr-x 0/0               0 2015-09-03 22:22 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/
-rw-r--r-- 0/0               3 2015-09-03 22:22 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/VERSION
-rw-r--r-- 0/0            1181 2015-09-03 22:22 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/json
-rw-r--r-- 0/0            1024 2015-09-03 22:22 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/layer.tar
-rw-r--r-- 0/0              90 2015-09-03 22:22 repositories
```

With this change, the layer's directory and artifact will have times
matching the image layer's created time. The "repositories" file is set
to epoch.

``` bash
vbatts@valse ~ (master) $ docker save busybox | tar tv
drwxr-xr-x 0/0               0 2015-04-17 18:01 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/
-rw-r--r-- 0/0               3 2015-04-17 18:01 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/VERSION
-rw-r--r-- 0/0            1405 2015-04-17 18:01 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/json
-rw-r--r-- 0/0         2643968 2015-04-17 18:01 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea/layer.tar
drwxr-xr-x 0/0               0 2015-04-17 18:01 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/
-rw-r--r-- 0/0               3 2015-04-17 18:01 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/VERSION
-rw-r--r-- 0/0            1346 2015-04-17 18:01 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/json
-rw-r--r-- 0/0            1024 2015-04-17 18:01 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55/layer.tar
drwxr-xr-x 0/0               0 2015-04-17 18:01 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/
-rw-r--r-- 0/0               3 2015-04-17 18:01 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/VERSION
-rw-r--r-- 0/0            1181 2015-04-17 18:01 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/json
-rw-r--r-- 0/0            1024 2015-04-17 18:01 cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff/layer.tar
-rw-r--r-- 0/0              90 1969-12-31 19:00 repositories
```

Side effect of this is that the tar stream from `docker save` is now
more deterministic.

``` bash
vbatts@valse ~ (master) $ docker save busybox | sha1sum
baf03e30ef79ca4d9c5e512d3a1b873880f404ca  -
vbatts@valse ~ (master) $ docker save busybox | sha1sum
baf03e30ef79ca4d9c5e512d3a1b873880f404ca  -
vbatts@valse ~ (master) $ docker save busybox | sha1sum
baf03e30ef79ca4d9c5e512d3a1b873880f404ca  -
vbatts@valse ~ (master) $ docker save busybox | sha1sum
baf03e30ef79ca4d9c5e512d3a1b873880f404ca  -
```

This idea came about from thinking about https://github.com/docker/docker/issues/8819#issuecomment-137606443

Signed-off-by: Vincent Batts vbatts@redhat.com

### CODE REVIEW

**Comment by tianon** [comment]
Wouldn't it make more sense to have `repositories` be set to the latest timestamp of the layers it represents?

**Comment by vbatts** [comment]
Maybe. There can be a number of images and tagged layers. Then it would be a sort and still variadic. Regardless of the arguments passed to `docker save` the time stamps will only be consistent in the way I've presently done this.

**Comment by calavera** [comment]
LGTM

**Comment by tianon** [comment]
@vbatts thanks for clarifying; makes sense :+1:

**Comment by vbatts** [comment]
I literally can't even lxc. No idea what these failures are, but w/e.

**Comment by cpuguy83** [comment]
How about an integration test?

**Comment by vbatts** [comment]
ooh

**Comment by vbatts** [comment]
@cpuguy83 done. PTAL

File: integration-cli/docker_cli_save_load_test.go [hunk 1]
```diff
@@ -94,6 +96,29 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
 }
 
+func (s *DockerSuite) TestSaveCheckTimes(c *check.C) {
+	repoName := "busybox:latest"
+	out, _ := dockerCmd(c, "inspect", repoName)
+	data := []struct {
+		ID      string
+		Created time.Time
+	}{}
+	if err := json.Unmarshal([]byte(out), &data); err != nil {
```
Comment (vdemeester) [inline integration-cli/docker_cli_save_load_test.go]: A nit, but :
```go
err := json.Unmarshal([]byte(out), &data)
c.Assert(err, checker.IsNil)
```

File: integration-cli/docker_cli_save_load_test.go [hunk 2]
```diff
@@ -94,6 +96,29 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+	}
+	if len(data) == 0 {
+		c.Fatalf("failed to marshal the data from %q", repoName)
+	}
+	tarTvTimeFormat := "2006-01-02 15:04"
+	out, _, err := runCommandPipelineWithOutput(
+		exec.Command(dockerBinary, "save", repoName),
+		exec.Command("tar", "tv"),
+		exec.Command("grep", "-E", fmt.Sprintf("%s %s", data[0].Created.Format(tarTvTimeFormat), data[0].ID)))
+	if err != nil {
```
Comment (vdemeester) [inline integration-cli/docker_cli_save_load_test.go]: Same here, could be :
```go
c.Assert(err, checker.IsNil, check.Commentf("failed to save repo with image ID and 'repositories' file: %s, %v", out, err))
```

File: integration-cli/docker_cli_save_load_test.go [hunk 3]
```diff
@@ -94,6 +96,29 @@ func (s *DockerSuite) TestSaveSingleTag(c *check.C) {
...
+	out, _ := dockerCmd(c, "inspect", repoName)
+	data := []struct {
+		ID      string
+		Created time.Time
+	}{}
+	if err := json.Unmarshal([]byte(out), &data); err != nil {
+		c.Fatalf("failed to marshal from %q: err", repoName, err)
+	}
+	if len(data) == 0 {
+		c.Fatalf("failed to marshal the data from %q", repoName)
```
Comment (vdemeester) [inline integration-cli/docker_cli_save_load_test.go]:
```go
c.Assert(len(data), checker.Equals, 0, check.Commentf("failed to marshal the data from %q", repoName))
```

**Comment by cpuguy83** [comment]
integration-cli/docker_cli_save_load_test.go:110: wrong number of args for format in Fatalf call: 1 needed but 2 args

**Comment by vbatts** [comment]
@cpuguy83 huh?
```bash
+ go test -check.f TestSaveCheckTimes -test.timeout=60m -check.v github.com/docker/docker/integration-cli
INFO: Testing against a local daemon
PASS: docker_cli_save_load_test.go:99: DockerSuite.TestSaveCheckTimes   0.062s
OK: 1 passed
PASS
```

File: integration-cli/docker_cli_save_load_test.go (hunk 2 above)
Comment (vbatts) [inline integration-cli/docker_cli_save_load_test.go]: even though this is copy-pasta from neighboring tests, i'll change it. ;-)

**Comment by vbatts** [comment]
@vdemeester updated to use assertion, though that loses the context message if it fails :-\

**Comment by vdemeester** [comment]
@vbatts why not using the `check.Commentf` thingy though ? :question:

**Comment by vbatts** [comment]
because
 
with this check package

**Comment by vbatts** [comment]
added commentf

**Comment by cpuguy83** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM, wating for janking :stuck_out_tongue_closed_eyes:

**Comment by vdemeester** [comment]
Same old windows failure `TestExecStartFails`, not related..

**Comment by paddie** [comment]
I'm attempting to do some local caching of separate layers of the build process for a particularly grueling npm install pass, and I'm hitting on an issue that seems related to this.

Using this minimal image:

```Docker
FROM alpine

CMD echo help
```

And running the following command on OSX repeatably, it produces different sha1 checksums.

I'm on OSX though, so I'm using the `openssl sha1` to compute the sha1:

```bash
> docker build -t help .
> docker save $(docker history -q help) | openssl sha1
35da520b07a54d5f771d9bdae6e8d646d99109e8
> docker save $(docker history -q help) | openssl sha1
530b5ce927101400c2c4c7ceb908e87108a0736a
```

The following command still produces consistent sha1:
```
> docker save help
```

Does anyone know what's the reason for the differences in the checksum that I'm seeing?

**Pull request merged**